### DISCUSSION TITLE
[bug] update_all does not map :as option of the field properly

### DISCUSSION THREAD

**Discussion by midnight-wonderer** [mongodb/mongoid#5666] [General]
Hi there
This is a bug report; please bear with me.

## Scenario

Supposed we have a simple model with as-option-set fields:
~~~ruby
class BuggyModel
  include ::Mongoid::Document
  store_in collection: 'buggy_models'

  field :_a, as: :alfa, type: :integer, default: 1
  field :_b, as: :bravo, type: :integer, default: 1

  # ignore the following field for now
  # it will be used in later arguments
  class CustomSerializer
    MAPPING = {
      'A' => 1,
      'B' => 2,
      'C' => 3,
    }.freeze

    INVERSE_MAPPING = MAPPING.invert.freeze

    class << self
      def mongoize(string)
        MAPPING[string]
      end

      def demongoize(integer)
        INVERSE_MAPPING[integer]
      end

      def evolve(string)
        mongoize(string)
      end
    end
  end
  field :charlie, type: CustomSerializer
end
~~~

I created one document in the collection:
~~~ruby
BuggyModel.create!({ _id: 'test' })
~~~

Then update the document:
~~~ruby
BuggyModel.where({ _id: 'test' }).update_all({ '$set': { alfa: 2 }, '$inc': { bravo: 1 } })
~~~

What result do you expect from the following code?
~~~ruby
BuggyModel.find('test').then{ |d| [d.alfa, d.bravo] }
~~~

`[2, 2]`? like me?
Nope, the current Mongoid returns `[1, 1]`.

Why?
Because `update_all` does not map the field name with `BuggyModel.aliased_fields` properly.

That's the problem.

## Supporting rationales

You may argue that this is working as intended.
Here is my argument against that hypothetical argument.

Aliased fields work inside criteria:
~~~ruby
# this work
BuggyModel.where({ alfa: 1 }).first
~~~

Fields with custom serializers work with `update_all`:
~~~ruby
# this work
BuggyModel.where({ _id: 'test' }).update_all({ '$set': { charlie: 'B' } })
# charlie field of the document will be set to 2 inside the database
# and will be mapped to B properly outside the database.
~~~

I can't think of a plausible explanation for why this should not work:
~~~ruby
# this does not work
BuggyModel.where({ _id: 'test' }).update_all({ '$set': { alfa: 2 } })
~~~

What do you guys say?

**Comment by jamis** [comment]
Thanks for the bug report! I've been looking into this one, and I might have a solution soon (need to discuss it with the team). You can follow the ticket here: https://jira.mongodb.org/browse/MONGOID-5632.

**Comment by alexbevi** [comment]
@midnight-wonderer just to close the loop on this discussion, the issue you described was fixed in Mongoid 8.0.6 and 8.1.2.
